### DISCUSSION TITLE
How to access files uploaded files outside of external grading?

### DISCUSSION THREAD

**Discussion by mfbutner** [PrairieLearn/PrairieLearn#6377] [Q&A] [answered]
How do you access files submitted via pl-file-upload inside of server.py? 

Right now I think the assumption is that if I'm using this element I must be using an external grader?

**Comment by nwalters512** [comment, marked as answer]
This is documented indirectly on the "Writing questions" section on this page: https://prairielearn.readthedocs.io/en/latest/externalGrading/#writing-questions. It's written for someone trying to implement their own file submissions mechanism, but you can apply this to reading files as well.

Basically, any files are stored in an array at `data["submitted_answers"]["_files"]`, and each element of the array consists of a `name` and base64-encoded `contents`.

See the implementation of `pl-file-preview` for an example of an element that reads submitted files:

https://github.com/PrairieLearn/PrairieLearn/blob/88379259e6fc79dd7c6a3d86352569308484ddd5/elements/pl-file-preview/pl-file-preview.py#L25-L44

**Reply by mfbutner** [reply]
Thanks! I don't think I ever would have thought to look there.
